Basic concepts Documentation and first Example (#139)
gardener/landscaper: docs/level_0/basic_concepts.md

# Basic Concepts

> This document describes the basic concepts, artifacts and their relationships of Landscaper. 

Reviewer: ```suggestion
This document describes the basic concepts, artifacts and their relationships within the Landscaper universe. 
```

## DeployItem

The main objective of Landscaper is to set up large cloud environments with complex relationships consisting of many

> small installation tasks like helm deployments, network setup etc. These elementary installation tasks are defined by 

Reviewer: ```suggestion
small installation tasks like Helm deployments, network setup etc. These elementary installation tasks are defined by 
```

> DeployItems. There are different types of DeployItems for example for deploying helm charts, plain kubernetes manifests 

Reviewer: ```suggestion
DeployItems. There are different types of DeployItems for example for deploying Helm charts, plain kubernetes manifests 
```

or terraform configurations.

> A DeployItem can be configurable such that it can be used in different scenarios. Typical examples of configurable parts

Reviewer: ```suggestion
A DeployItem can be configured such that it can be used in different scenarios. Typical examples of configurable parts
```

of a DeployItem are the target cluster or namespace where a helm chart should be deployed to. A DeployItem can also 

> define output parameter for data it creates and which can be consumed by others.

Reviewer: ```suggestion
define output parameters for data it creates and which can be consumed by others.
```

## Blueprint

> Blueprints are reusable collections of combined installation tasks which could be parameterized through import
> parameters. They describe system setups consisting of a set of installations task and their dependencies.

Reviewer 2: ```suggestion
Blueprints are reusable collections of combined installation tasks which could be parameterized through import
parameters.
```
the same information is already said as `reusable collections of combined installation tasks`. 
alternativly we could have something like:
`Blueprints are reusable system setups which could be parameterized through import
parameters. System setups are collections of combined installations task and their dependencies.`

> Several DeployItems can be defined in a Blueprint. By this, a Blueprint becomes a set of installation tasks. 
> 
> A Blueprint defines an interface for import data, required by its DeployItems. It can also define an export 
> interface to expose output data of its DeployItems. Furthermore, the execution order of the DeployItems of a Blueprint 
> can be specified.

Reviewer 2: ```suggestion
Several DeployItems can be defined in a Blueprint. 
Also the execution order of the DeployItems of a Blueprint can be specified.

A Blueprint defines an interface for import data, required by its DeployItems. It can also define an export 
interface to expose output data of its DeployItems. 
```

## Blueprint Component and Component Descriptor

A component is a quite general term usually describing some IT artifact ranging from a small program to large
and complex systems. In the context of Landscaper we focus on components setting up some cloud environment 
described with a Blueprint. This could range from quite simple installations of one or two helm charts, or the

> setup up of a complex system like a Garden landscape.

Reviewer 2: ```suggestion
setup up of a complex system like a Gardener landscape.
```

> A Blueprint Component usually consists of a Blueprint, and the resources required by the DeployItems, like helm charts, docker images, json schema etc. All these required resources are described by a Component 

Reviewer: ```suggestion
A Blueprint Component usually consists of a Blueprint and the resources required by the DeployItems, like Helm charts, Docker images, JSON schema etc. All these required resources are described by a Component 
```

Descriptor as a yaml file. 

Component Descriptors are a complete description of all resources belonging to a component and could be used for
different tasks like security scanning or transport.

Usually Component Descriptors and Blueprints are stored in an OCI registry. More details about Component 
Descriptors and Component Descriptor Artifacts can be found [here](https://github.com/gardener/component-spec).

## Installation

> A Blueprints is a reusable definition of the installation process of a particular cloud environment. 

Reviewer: ```suggestion
A Blueprint is a reusable definition of the installation process of a particular component. 
```

Reviewer: Cloud environment or component?

An instance of such a Blueprint with particular import data is called an Installation. Installations are Kubernetes 
Custom Resources (CR) defined by Landscaper. An installation references the Blueprint of a Component via the 
corresponding Component Descriptor. It specifies the import data and how to handle the export data. 

> An Installation is deployed to the cluster where Landscaper watches these CRs. If Landscaper recognizes an installation

Reviewer: ```suggestion
An Installation is deployed to the cluster where the landscaper-controller watches these CRs. If it recognizes an installation
```

# Basic Concepts

This document describes the basic concepts, artifacts and their relationships of Landscaper. 

## DeployItem

The main objective of Landscaper is to set up large cloud environments with complex relationships consisting of many
small installation tasks like helm deployments, network setup etc. These elementary installation tasks are defined by 
DeployItems. There are different types of DeployItems for example for deploying helm charts, plain kubernetes manifests 

> or terraform configurations.

Reviewer 3: Shouldn't we have a dedicated "Basic Concepts" Chapter, and not start with "DeployItems" immediately ?

Author: What should be the content of such a "Basic Concepts" Chapter? If you go to the [overview](https://github.com/robertgraeff/landscaper/blob/basic-concepts-2/docs/README_new.md) you see a motivation chapter which should contain the even more basic details.

A Blueprint can define an interface for import data, required by its DeployItems. It can also define an export 
interface to expose output data of its DeployItems. Furthermore, the execution order of the DeployItems of a Blueprint 
could be specified. 

Blueprints are reusable collections of combined installation tasks which could be parameterized through their import 
specification/parameters. They describe system setups consisting of a set of installations task and their dependencies. 

## Blueprint Component and Component Descriptor

A component is a quite general term usually describing some IT artifact ranging from a small program to large
and complex systems. In the context of Landscaper we focus on components setting up some cloud environment 
described with a Blueprint. This could range from quite simple installations of one or two helm charts, or the
setup up of a complex system like a Garden landscape.

A Blueprint Component usually consists of a Blueprint, and the resources required by the DeployItems included in the 
Blueprint, like helm charts, docker images, json schema etc. All these required resources are described by a Component 
Descriptor as a yaml file. 

> Component Descriptors are a complete description of all stuff belonging to a component and could be used for

Reviewer 4: "of all stuff" sounds rather informal

Author: we have removed this

More details about Component Descriptors and Component Descriptor Artifacts can be found 
[here](https://github.com/gardener/component-spec).

## Installation

Blueprints, Component Descriptors, and Component Archives do not trigger any installation process. A Blueprints is a reusable 
definition of the installation process of a particular cloud environment. An instance of such a Blueprint with particular
import data is called an Installation. Installations are kubernetes Custom Resources (CR) defined by Landscaper.
An installation references the Blueprint of a Component via the corresponding Component Archive.
It specifies the import data and how to handle the export data. 

An Installation is deployed to the cluster where Landscaper watches these CRs. If Landscaper recognizes an installation
with all input data available, it starts the setup specified in the Blueprint. This executes the installation tasks 
defined in the DeployItems.

## Aggregated Blueprint

Very complex systems require hundreds of installation steps resulting in Blueprints with many DeployItems. 
Landscaper allows defining subsystems each described by a separate Blueprint and to combine these in an aggregated 

> Blueprint. An aggregated Blueprint references the Blueprints of the subsystems not directly but via Sub-Installations. 

Reviewer 4: Could be helpful to reintroduce Sub-Installations as an instance of the "separate Blueprint" of a subsystem. Although this is described in the previous paragraph, it makes it simpler to understand at this point.

Author: We described sub installations in more detail

# Basic Concepts

This document describes the basic concepts, artifacts and their relationships of Landscaper. 

## DeployItem

The main objective of Landscaper is to set up large cloud environments with complex relationships consisting of many
small installation tasks like helm deployments, network setup etc. These elementary installation tasks are defined by 
DeployItems. There are different types of DeployItems for example for deploying helm charts, plain kubernetes manifests 
or terraform configurations.

A DeployItem can define import parameter for data needed for the installation task, e.g. the target cluster where 
a helm chart should be deployed. A DeployItem can also define output parameter for data it creates and which can be 

> consumed by others.

Author: That's true. We will change this.

## Blueprint

Several DeployItems can be collected in a Blueprint. In this sense, a Blueprint is a set of installation tasks. 

A Blueprint can define an interface for import data, required by its DeployItems. It can also define an export 
interface to expose output data of its DeployItems. Furthermore, the execution order of the DeployItems of a Blueprint 

> could be specified. 

Reviewer 2: ```suggestion
can be specified. 
```

Blueprints are reusable collections of combined installation tasks which could be parameterized through their import 

> specification/parameters. They describe system setups consisting of a set of installations task and their dependencies. 

Reviewer 2: this should be at the beginning of the blueprints section.
So that the reader gets a high level explenation of blueprints and then we can describe the blueprint in more detail with deployitems etc

Author: we have changed this

Blueprint, like helm charts, docker images, json schema etc. All these required resources are described by a Component 
Descriptor as a yaml file. 

Component Descriptors are a complete description of all resources belonging to a component and could be used for
different tasks like security scanning or transport.

## Component Archive

Usually Component Descriptors are stored in Component Archive in an OCI registry. A Component Archive 
contains a Component Descriptor and optionally some other resources. Be aware that not all resources
of a component are stored in one Component Archive. The Component Descriptor can also just contain
references to resources located somewhere else, like a helm chart in some remote helm chart repository.

Blueprints are also stored in an OCI registry. They can be stored as a part of the Component Archive 
they belong to. Alternatively, they can be stored as a standalone OCI artifact referenced by the Component Descriptor of
a Component Archive.

More details about Component Descriptors and Component Descriptor Artifacts can be found 
[here](https://github.com/gardener/component-spec).

## Installation

> Blueprints, Component Descriptors, and Component Archives do not trigger any installation process. A Blueprints is a reusable 

Reviewer 2: ```suggestion
A Blueprint is a reusable 
```

> definition of the installation process of a particular cloud environment. An instance of such a Blueprint with particular

Reviewer 2: ```suggestion
definition of the installation process of a particular component/application. An instance of such a Blueprint with particular
```
a blueprint can describe a cloud environment but I think application/component/software is better suited here

> import data is called an Installation. Installations are kubernetes Custom Resources (CR) defined by Landscaper.

Reviewer 2: ```suggestion
import data is called an Installation. Installations are Kubernetes Custom Resources (CR).
```

> An installation references the Blueprint of a Component via the corresponding Component Archive.

Reviewer 2: here you use component and not blueprint component. I think we should stick to one term

Reviewer 2: If we mention the component descriptor we should explicitly say that blueprints are artifacts that are referenced in a component descriptor.

> It specifies the import data and how to handle the export data. 

Reviewer 2: that sentence seems bit out of context?

An Installation is deployed to the cluster where Landscaper watches these CRs. If Landscaper recognizes an installation

> with all input data available, it starts the setup specified in the Blueprint. This executes the installation tasks 
> defined in the DeployItems.

Reviewer 2: ```suggestion
with all input data available, it starts the setup specified in the Blueprint. This executes the installation tasks 
defined in the DeployItems.
```
i think we should either remove `This executes the installation tasks 
defined in the DeployItems.` or go into more detail.

## Aggregated Blueprint

Very complex systems require hundreds of installation steps resulting in Blueprints with many DeployItems. 
Landscaper allows defining subsystems each described by a separate Blueprint and to combine these in an aggregated 

> Blueprint. An aggregated Blueprint references the Blueprints of the subsystems not directly but via Sub-Installations.

Reviewer 2: what does `not directly` mean?

> Sub-Installations are again Installations referencing a Blueprint, which describes the subsystem.

Reviewer 2: in an aggregated blueprint you basically glue together multiple blueprint.
You do this by definfing instalaltion templates.
Theses installation templates will result in subinstallation but that might be to confusing...

Author: We think both sub-installations and installation templates are more confusing than helpful at this level. Its better to introduce them later. We will remove Sub-Installations and only explain the possiblity to combine blueprints in an aggregated blueprint without further details here.

Assume an Installation referencing an aggregated Blueprint is deployed. When all input data of the Installation is 

> available, all Sub-Installations are created. Sub-Installations get their input data from the parent installation as well

Reviewer 2: `When all input data of the Installation is 
available` - already said that in the last section

Author: We have restructured this paragraph.

as from the export data of their sibling Installations. Execution of a Sub-Installation is started, when all its import

> data is available. 

Reviewer 2: sub installations can use the imports defined by their parent or they can use the export of other subinstallations.
I think we have to explain that in more detail. Otherwise is seems confusing.

Author: We will add that sub installations get their data from the **input data** of the parent installation as well as the exports of siblings. We think that should be sufficient here. More explanations should follow e.g. in the tutorials.

> The hierarchy of installations/blueprints can be arbitrarily deep but in practice two levels should be sufficient.

Reviewer 2: ```suggestion
The hierarchy of installations/blueprints can be arbitrarily deep but in practice two levels should be sufficient.
```
why are 2 levels sufficient?

Author: we have removed this
